File: python/bundled_agents/io/io_bridge/src/io_bridge/egress_rules/password.py
"""`password` — kernel-GROUP membership by a shared secret (egress side: PRESENT)."""

from __future__ import annotations

import os
from dataclasses import dataclass

from io_bridge._base import EgressRule


@dataclass(frozen=True)
class Password(EgressRule):
    """Present this leg's group token (read from `token_env`, default
    `FANTASTIC_GROUP_TOKEN`) on every outbound `call`, so a paired group member's
    ingress `password` rule accepts it. The symmetric mirror of
    `ingress_rules.password`. Presents nothing when the env var is unset/empty."""

    token_env: str = "FANTASTIC_GROUP_TOKEN"

    def credential(self) -> str | None:
        return os.environ.get(self.token_env) or None

File: src/lib/rust/crates/fantastic-io-bridge/src/authorizer/egress/password.rs
//! `password` — kernel-GROUP membership by a shared secret (egress side: PRESENT).

use super::super::{EgressRule, DEFAULT_TOKEN_ENV};

/// Present this leg's group token (read from `token_env`, default
/// `FANTASTIC_GROUP_TOKEN`) on every outbound `call`, so a paired group member's
/// ingress `password` rule accepts it. The symmetric mirror of `ingress::password`.
/// Presents nothing when the env var is unset/empty.
pub struct Password {
    token_env: String,
}

impl Password {
    /// Build from an optional `token_env` (default `FANTASTIC_GROUP_TOKEN`).
    pub fn new(token_env: Option<String>) -> Self {
        Self {
            token_env: token_env.unwrap_or_else(|| DEFAULT_TOKEN_ENV.to_string()),
        }
    }
}

impl EgressRule for Password {
    fn credential(&self) -> Option<String> {
        std::env::var(&self.token_env)
            .ok()
            .filter(|s| !s.is_empty())
    }
}
